fw/rbcx-coprocessor/src/CdcUartTunnel.cpp

```diff
@@ -0,0 +1,99 @@
+/// Implements passthrough between the "primary" UART and USB CDC.
+
+#include "stm32f1xx_hal.h"
+#include "stm32f1xx_hal_dma.h"
+#include "stm32f1xx_ll_usart.h"
+
+#include "Bsp.hpp"
+#include "ByteFifo.hpp"
+#include "CdcUartTunnel.hpp"
+#include "UsbCdcLink.h"
+
+DMA_HandleTypeDef primaryUartDmaRxHandle;
+DMA_HandleTypeDef primaryUartDmaTxHandle;
+ByteFifo<512> primaryUartRxFifo;
+std::array<uint8_t, CDC_DATA_SZ> primaryUartTxBuf;
+
+void primaryUartInit() {
+    __HAL_RCC_USART1_CLK_ENABLE();
+    __HAL_RCC_DMA1_CLK_ENABLE();
+
+    LL_USART_InitTypeDef init;
+    LL_USART_StructInit(&init);
+    init.BaudRate = 115200;
+    init.DataWidth = LL_USART_DATAWIDTH_8B;
+    init.HardwareFlowControl = LL_USART_HWCONTROL_NONE;
+    init.Parity = LL_USART_PARITY_NONE;
+    init.StopBits = LL_USART_STOPBITS_1;
+    init.TransferDirection = LL_USART_DIRECTION_TX_RX;
+    LL_USART_Init(USART1, &init);
+    LL_USART_Enable(USART1);
+
+    // UART RX runs indefinitely in circular mode
+    primaryUartDmaRxHandle.Instance = DMA1_Channel5;
+    primaryUartDmaRxHandle.Init.Direction = DMA_PERIPH_TO_MEMORY;
+    primaryUartDmaRxHandle.Init.Mode = DMA_CIRCULAR;
+    primaryUartDmaRxHandle.Init.MemInc = DMA_MINC_ENABLE;
+    primaryUartDmaRxHandle.Init.PeriphInc = DMA_PINC_DISABLE;
+    primaryUartDmaRxHandle.Init.MemDataAlignment = DMA_MDATAALIGN_BYTE;
+    primaryUartDmaRxHandle.Init.PeriphDataAlignment = DMA_PDATAALIGN_BYTE;
+    primaryUartDmaRxHandle.Init.Priority = DMA_PRIORITY_MEDIUM;
+    HAL_DMA_Init(&primaryUartDmaRxHandle);
+    HAL_DMA_Start(&primaryUartDmaRxHandle, uint32_t(&(USART1->DR)), uint32_t(primaryUartRxFifo.data()), primaryUartRxFifo.size());
+    LL_USART_EnableDMAReq_RX(USART1);
+
+    // UART TX burst is started ad hoc each time
+    primaryUartDmaTxHandle.Instance = DMA1_Channel4;
+    primaryUartDmaTxHandle.Init.Direction = DMA_MEMORY_TO_PERIPH;
+    primaryUartDmaTxHandle.Init.Mode = DMA_NORMAL;
+    primaryUartDmaTxHandle.Init.MemInc = DMA_MINC_ENABLE;
+    primaryUartDmaTxHandle.Init.PeriphInc = DMA_PINC_DISABLE;
+    primaryUartDmaTxHandle.Init.MemDataAlignment = DMA_MDATAALIGN_BYTE;
+    primaryUartDmaTxHandle.Init.PeriphDataAlignment = DMA_PDATAALIGN_BYTE;
+    primaryUartDmaTxHandle.Init.Priority = DMA_PRIORITY_MEDIUM;
+    HAL_DMA_Init(&primaryUartDmaTxHandle);
+    LL_USART_EnableDMAReq_TX(USART1);
+
+    __HAL_RCC_GPIOA_CLK_ENABLE();
+    pinInit(GPIOA, GPIO_PIN_9, GPIO_MODE_AF_PP, GPIO_PULLUP, GPIO_SPEED_FREQ_HIGH);
+    pinInit(GPIOA, GPIO_PIN_10, GPIO_MODE_AF_INPUT, GPIO_PULLUP, GPIO_SPEED_FREQ_HIGH);
```

Comment:
Tady se vyskytují konkrétní piny - neměla by jejich definice být v BSP, aby skutečně byla zachována portovatelnost mezi deskami? Stejně tak přemýšlím nad DMA kanály, ale ty budou asi i mezi rozdílnými MCU stejné, že? STM32G totiž mají DMA mux a je jedno, který kanál použiješ na co.

Reply:
DMA přiřazení je mezi F1 totožné. V BSP by piny měly být. Sám jsem to viděl jako tolerovatelné, ale spravím to :)

Reply:
Ono je to takový na prd. Závislost na konkrétním USARTu jsou:
- piny
- instance USARTu
- DMA kanály
- __HAL_RCC_USART1_CLK_ENABLE
- __HAL_RCC_DMA1_CLK_ENABLE

Reply:
Provedeno c5ed08b